fw/http_msg.c

```diff
@@ -815,6 +820,33 @@ __hdr_sub(TfwHttpMsg *hm, const TfwStr *hdr, unsigned int hid)
 	return TFW_PASS;
 }
 
+/**
+ * Substitute request method with "get"
+ *
+ * Note: replaced method name length is supposed to be shorter
+ *       than the replacement string ("get")
+ */
+int
+tfw_http_req_meth_sub_with_get(TfwHttpMsg *hm, unsigned char dest_meth)
+{
+	const char *dest_name;
+	size_t dest_len;
+
+	TfwStr sub_str = TFW_STR_STRING("get");
+
+	BUG_ON(hm->req->method != dest_meth);
+
+	dest_name = tfw_http_tbl_method_get_name(dest_meth);
+	dest_len = strlen(dest_name);
+
+	BUG_ON(dest_len < sub_str.len);
+
+	if (ss_skb_cutoff_data(hm->msg.skb_head, &sub_str, 0, dest_len-sub_str.len))
+		return TFW_BLOCK;
```

Comment:
Just an idea: perhaps, a smarter modification may be considered that would require less modification of the underlying SKB (and thus be faster). In case of HTTP/1.1 this header line is the first line in a request. It can be argued that in many cases it would be the start of SKB data. If we align the new string to the right of the previous method's string (`PURGE` -> PU**GET**), then the modification of the SKB would be a simple adjustment of pointers (move by two chars forward) instead of a heavier and more expensive modification in the middle of the SKB.

Not sure if this trick is applicable to HTTP/2.  Perhaps not.

Reply:
for a while I didn't realize this. but I like the idea, I will consider it for further redesign.

Reply:
Yeah, the @keshonok 's proposal is the right way to go. We actuall need to do this in two steps (in any order):
1. call the skb cutting logic (IIRC it's already optimized for the case, when the beginning is cut)
2. rewrite the 3 bytes